pkg/querier/seriesset.go

```diff
@@ -0,0 +1,244 @@
+
+	if s.aggrSeries != nil && s.aggrSeries.CanAggregate(s.partition.AggrType()) && s.maxt-s.mint >= s.interval {
+		s.attrs = s.aggrSeries.GetAttrNames()
+	} else {
+		s.attrs, s.chunkIds = s.partition.Range2Attrs("v", s.mint, s.maxt)
+	}
+	attrs = append(attrs, s.attrs...)
+
+	s.logger.DebugWith("Select - GetItems", "path", path, "attr", attrs, "filter", filter, "name", name)
+	input := v3io.GetItemsInput{Path: path, AttributeNames: attrs, Filter: filter, ShardingKey: name}
+	iter, err := utils.NewAsyncItemsCursor(container, &input, workers, shardingKeys)
+	if err != nil {
+		return err
+	}
+
+	s.iter = iter
+	return nil
+
+}
+
+// advance to the next series
+func (s *V3ioSeriesSet) Next() bool {
+
+	// create raw chunks series (not aggregated)
+	if s.aggrSeries == nil {
+		if s.iter.Next() {
+			s.currSeries = NewSeries(s)
+			return true
+		}
+		return false
+	}
+
+	// create multiple aggregation series (one per aggregation function)
+	// the index is initialized as numfunc-1 (so the first +1 and modulo will be eq 0)
+	if s.aggrIdx == s.aggrSeries.NumFunctions()-1 {
+		// if no more items (from GetItems cursor), return with EOF
+		if !s.iter.Next() {
+			return false
+		}
+
+		s.nullSeries = false
+
+		if s.aggrSeries.CanAggregate(s.partition.AggrType()) && s.maxt-s.mint > s.interval {
+
+			// create series from aggregation arrays (in DB) if the partition stored the desired aggregates
+			maxtUpdate := s.maxt
+			maxTime := s.iter.GetField("_maxtime")
+			if maxTime != nil && int64(maxTime.(int)) < s.maxt {
+				maxtUpdate = int64(maxTime.(int))
+			}
+			mint := s.partition.CyclicMinTime(s.mint, maxtUpdate)
+
+			start := s.partition.Time2Bucket(mint)
+			end := s.partition.Time2Bucket(s.maxt + s.interval)
+
+			// len of the returned array, cropped at the end in case of cyclic overlap
+			length := int((maxtUpdate-mint)/s.interval) + 2
+
+			if s.overlapWin != nil {
+				s.baseTime = s.maxt //- int64(s.overlapWin[0]) * s.interval
+			} else {
+				s.baseTime = mint
+			}
+
+			if length > 0 {
+				attrs := s.iter.GetFields()
+				aggrSet, err := s.aggrSeries.NewSetFromAttrs(length, start, end, mint, s.maxt, &attrs)
+				if err != nil {
+					s.err = err
+					return false
+				}
+
+				s.aggrSet = aggrSet
+			} else {
+				s.nullSeries = true
+			}
+
+		} else {
+
+			// create series from raw chunks
+			s.currSeries = NewSeries(s)
+			s.aggrSet = s.aggrSeries.NewSetFromChunks(int((s.maxt-s.mint)/s.interval) + 1)
+			if s.overlapWin != nil {
+				s.chunks2WindowedAggregates()
+			} else {
+				s.chunks2IntervalAggregates()
+			}
+
+		}
+	}
+
+	s.aggrIdx = (s.aggrIdx + 1) % s.aggrSeries.NumFunctions()
+	return true
+}
+
+// convert raw chunks to fixed interval aggregator
+func (s *V3ioSeriesSet) chunks2IntervalAggregates() {
+
+	iter := s.currSeries.Iterator()
+	if iter.Next() {
+
+		if iter.Err() != nil {
+			s.err = iter.Err()
+			return
+		}
+
+		t0, _ := iter.At()
+		s.baseTime = (t0 / s.interval) * s.interval
+
+		for {
```

Comment:
isn't it equivalent to `for ; iter.Next(); {...}`

Reply:
@imakhlin no, we need to init baseTime, but there seem to be a bug (the if iter.Err() need to go out of the if, since .Next will return false in case of error)